subsys/bluetooth/host/l2cap.c

```diff
@@ -2392,6 +2497,49 @@ static void l2cap_chan_le_recv_seg(struct bt_l2cap_le_chan *chan,
 	l2cap_chan_le_recv_sdu(chan, buf, seg);
 }
 
+#ifdef CONFIG_BT_L2CAP_SEG_RECV
+static void l2cap_chan_le_recv_seg_direct(struct bt_l2cap_le_chan *chan, struct net_buf *seg)
+{
+	uint16_t seg_offset;
+	uint16_t sdu_remaining;
+
+	if (chan->_sdu_len_done == chan->_sdu_len) {
+
+		/* This is the first PDU in a SDU. */
+
+		if (seg->len < 2) {
+			LOG_WRN("Missing SDU header");
+			bt_l2cap_chan_disconnect(&chan->chan);
+			return;
+		}
+
+		/* Pop off the "SDU header". */
+		chan->_sdu_len = net_buf_pull_le16(seg);
+		chan->_sdu_len_done = 0;
+
+		if (chan->_sdu_len > chan->rx.mtu) {
+			LOG_WRN("SDU exceeds MTU");
+			bt_l2cap_chan_disconnect(&chan->chan);
+			return;
+		}
+	}
+
+	seg_offset = chan->_sdu_len_done;
+	sdu_remaining = chan->_sdu_len - chan->_sdu_len_done;
+
+	if (seg->len > sdu_remaining) {
+		LOG_WRN("L2CAP RX PDU total exceeds SDU");
+		bt_l2cap_chan_disconnect(&chan->chan);
+	}
+
+	/* Commit receive. */
+	chan->_sdu_len_done += seg->len;
+
+	/* Tail call. */
+	chan->chan.ops->seg_recv(&chan->chan, chan->_sdu_len, seg_offset, seg->len, seg->data);
```

Comment:
I'm wondering if we could pass the `net_buf_simple` part of `seg`. That way the app has access to the net/buf.h convenience fns (pull_mem, pull various sizes etc).
what do you think?

Reply:
The app can itself instantiate a `net_buf_simple` from this buffer. A benefit of passing a `net_buf_simple` would be that the app can use the whole stack ACL buffer if it wants to. 

I found it both hard to document `seg_recv` when it does not get the buffer length as a parameter and hard to document that the buffer backing the net_buf_simple is ok-to-clobber, but its lifetime ends when this function returns.